### PR TITLE
Add Fusion-Scalar transformation to SCA

### PR DESCRIPTION
**Pull request by MDosky**
## Status

**REVIEW**

## Description
Use of loop fusion and keeping of scalars if possible.
It works differently from the other version since it keep track of the variable use in "blocks". It than proceed to use the various sets to know how to handle promotions.

### CODE REVIEW

**Comment by MDosky** [comment]
@clementval This should be 'reviewable' for now.

**Comment by MDosky** [comment]
Results / Evident bugs:
 sca23/24: assignment of `res = t` which are vectors are grouped inside the last `DO` loop. Thinking about a solution.
sca34/35: `IF` is outside the `DO` statement, should be easily solvable.
sca36: `p` isn't promoted correctly, but since it's influenced by `t` which is promoted there is a useless loop around it in the beginning.

File: cx2t/src/claw/wani/transformation/sca/Parallelize.java [hunk 1]
```diff
@@ -472,6 +476,252 @@ private void transformForGPU(XcodeProgram xcodeml)
...
+    // Keep track of the block we are in
+    int depth = 0;
+
+    // Extract all the variables used, this doesn't include vector's
+    // iterator variables
+    transformPartialForCPUAnalysis(_fctDef.body(), depth, depthVars,
+            affectingVars);
+
+    // Find the block we need to transform
+    Set<String>[] targetDepthIntersections = new Set[depthVars.size()];
```
Comment (MDosky) [inline cx2t/src/claw/wani/transformation/sca/Parallelize.java]: Array of generic are a bad idea, or 'unsafe' as the compiler put it.
Should be replaced by a List.

**Comment by clementval** [comment]
@MDosky Would be better if the PR can pass the test... If any reference needs to be updated you can update them. This will be part of the review as well. 
By the way, I'll try to review what you have done so far until the meeting tomorrow.

File: cx2t/src/claw/wani/transformation/sca/Parallelize.java (hunk 1 above)
Comment (clementval) [reply cx2t/src/claw/wani/transformation/sca/Parallelize.java]: If so, change it directly.

File: cx2t/src/claw/wani/transformation/sca/Parallelize.java [hunk 2]
```diff
@@ -472,6 +476,251 @@ private void transformForGPU(XcodeProgram xcodeml)
...
+
+  /**
+   * Given a series of sequential hooks (nodes), envelope them in a DO loop.
+   *
+   * @param hooks List of nodes do envelope in a DO statement
+   * @param xcodeml The program
+   */
+  private void transformPartialForCPUApply(List<Xnode> hooks,
+                                             XcodeProgram xcodeml) {
+    if (hooks.isEmpty()) return;
```
Comment (clementval) [inline cx2t/src/claw/wani/transformation/sca/Parallelize.java]: Missing {}

File: cx2t/src/claw/wani/transformation/sca/Parallelize.java [hunk 3]
```diff
@@ -472,6 +476,252 @@ private void transformForGPU(XcodeProgram xcodeml)
     Directive.generateRoutineDirectives(xcodeml, _fctDef);
   }
 
+  /**
+   * Apply partial CPU transformation.
+   * When we can fusion the loop and keep the scalar "as is", we will do that
```
[the remark this answers is not in the document]
Comment (MDosky) [reply cx2t/src/claw/wani/transformation/sca/Parallelize.java]: Yes, it also wasn't updated from the previous version.

**Comment by MDosky** [comment]
sca23, sca24 and sca36 should still fail after the reference update for now.

**Comment by MDosky** [comment]
@clementval it should be OK from commit eed72db, but now travis seems to have a problem with gcc-6.

File: test/claw/sca/sca20/reference_cpu.f90 [hunk 4]
```diff
@@ -17,8 +17,6 @@ FUNCTION compute_column ( nz , b , q , t , nproma , o ) RESULT(r)
   DO k = 2 , nz , 1
    DO proma = 1 , nproma , 1
     t ( proma , k ) = c * k
-   END DO
```
Comment (clementval) [inline test/claw/sca/sca20/reference_cpu.f90]: Ok

File: test/claw/sca/sca21/reference_cpu.f90 [hunk 5]
```diff
@@ -17,8 +17,6 @@ FUNCTION compute_column ( nz , b , q , t , nproma , o ) RESULT(r)
   DO k = 2 , nz , 1
    DO proma = 1 , nproma , 1
     t ( proma , k ) = c * k
-   END DO
```
Comment (clementval) [inline test/claw/sca/sca21/reference_cpu.f90]: Ok

File: test/claw/sca/sca22/reference_cpu.f90 [hunk 6]
```diff
@@ -35,8 +35,6 @@ SUBROUTINE compute_column ( nz , q , t , nproma )
   DO k = 2 , nz , 1
    DO proma = 1 , nproma , 1
     t ( k , proma ) = c * k
-   END DO
```
Comment (clementval) [inline test/claw/sca/sca22/reference_cpu.f90]: Ok

File: test/claw/sca/sca23/reference_cpu.f90 [hunk 7]
```diff
@@ -15,8 +15,6 @@ FUNCTION compute_column ( nz , q , t , nproma ) RESULT(res)
   DO k = 2 , nz , 1
    DO proma = 1 , nproma , 1
     t ( k , proma ) = c * k
-   END DO
```
Comment (clementval) [inline test/claw/sca/sca23/reference_cpu.f90]: ok

File: cx2t/src/claw/tatsu/xcodeml/xnode/XnodeUtil.java [hunk 8]
```diff
@@ -410,6 +410,25 @@ private static NodeList evaluateXpath(Element from, String xpath)
     return realReferences;
   }
 
+  /**
+   * Find all Xnode.VAR inside the given node and return their value. From
+   * the set are excluded the variables used as indexes for vectors.
+   *
+   * @return A set contains the variables used inside the node.
+   */
+  public static Set<String> findChildrenVariables(Xnode node) {
```
[the remark this answers is not in the document]
Comment (MDosky) [reply cx2t/src/claw/tatsu/xcodeml/xnode/XnodeUtil.java]: But it's in `claw.tatsu.xcodeml.xnode.XnodeUtil`, did you mean somewhere different?

[the remark this answers is not in the document]
Comment (clementval) [reply cx2t/src/claw/tatsu/xcodeml/xnode/XnodeUtil.java]: My bad, it is indeed in the correct class.

**Comment by clementval** [comment]
Let's keep only the branch one

**Pull request closed without merging**